apps/05Mastermind.js

```diff
@@ -26,27 +28,64 @@ function getRandomInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-    // your code here
+function generateHint(solution, gess) {
+
+    var solutionArray = [];
+    var guessArray = [];
+    var correctLetterLocations = 0;
+    var correctLetters = 0;
+    var targetIndex = 0;
+
+    solutionArray = solution.split("");
+    guessArray = gess.split("");
+    for (var i=0; i<4;i++){
+      if (guessArray[i] === solutionArray[i]){
+        correctLetterLocations++;
+        solutionArray[i] = null;
+        guessArray[i] = '';
+      }
+    }
+    for (var i=0; i<4;i++){
+      targetIndex = solutionArray.indexOf(guessArray[i]);
+      if (targetIndex>-1){
+          correctLetters ++;
+          solutionArray[targetIndex] = null;
+      }
+    }
+    return correctLetterLocations+"-"+correctLetters;
 }
 
 function mastermind(guess) {
-    // your code here
+    var win = "";
+    if (guess === solution) {
+      win = "You guessed it!" ;
+    } else {
+      board[board.length] = guess;
+      win = generateHint(solution, guess);
+    }
+    return win;
 }
 
-
 function getPrompt() {
     prompt.get(['guess'], function (error, result) {
-        console.log( mastermind(result['guess']) );
-        printBoard();
-        getPrompt();
+        console.log(solution) ;
+        hint = mastermind(result['guess']);
+        console.log(hint) ;
+        if (board.length>=10) {
+          console.log('You ran out of turns! The solution was '+solution);
+        } else {
+          if (!(hint==='You guessed it!')) {
```

Comment:
This looks like it works but let me say this: it seems like we're checking "did we win" which we already checked up on line 60. This is a good indication that we're duplicating logic and we could consolidate